### PR TITLE
Adding Highway Network layers (regular and time-distr.)

### PR DESCRIPTION
**Pull request by lukedeo**
@fchollet let me know what you think. In particular, lemme know if theres anything I'm doing that breaks the doc autogen.

### CODE REVIEW

File: keras/layers/core.py [hunk 1]
```diff
@@ -1694,3 +1795,92 @@ def add_shared_layer(layer, inputs):
...
+            self.set_weights(self.initial_weights)
+            del self.initial_weights
+
+    @property
+    def output_shape(self):
+        return (self.input_shape[0], self.input_shape[1])
+
+    def get_output(self, train=False):
+        X = self.get_input(train)
+        transform_weight = activations.sigmoid(K.dot(X, self.W_carry) + self.b_carry)
```
Comment (fchollet) [inline keras/layers/core.py]: Shouldn't the inner activation be configurable instead of being hard-coded to sigmoid?

**Comment by fchollet** [comment]
So in your experience, how well would a stack of highway units perform against a stack of dense units (with dropout regularization)?

File: keras/layers/core.py (hunk 1 above)
Comment (lukedeo) [inline keras/layers/core.py]: In general I suppose so, but there is the problem of `(a, (1-a))` needing to define a simplex. In the paper, they use sigmoid. Think it would be useful to allow people to use `hard_sigmoid` or `ultra_fast_sigmoid`?

**Comment by lukedeo** [comment]
In general, I've found they converge a bit faster, and can be made *much* deeper, but to be honest I've found them *most* useful when tacked on to RNNs.

**Comment by lukedeo** [comment]
(unrelated thing causing travis to fail...)

**Comment by lukedeo** [comment]
@fchollet or @farizrahman4u, do you have any insights as to why unrelated test seem to be failing?

**Comment by farizrahman4u** [comment]
@lukedeo Its alright. Its just a flaky test. The final loss in that test is kind of random and it depends on many things like numpy's random number generator etc..it goes off from time to time.

**Comment by lukedeo** [comment]
Yeah, it's a bit disconcerting. Do you have any methods to get it to pass so the PR can be merged?

**Comment by fchollet** [comment]
Restarting the tests.

**Comment by lukedeo** [comment]
Dope, thanks.

**Comment by fchollet** [comment]
The `Highway` layer will be great to have. I'm wondering wether we should merge the TimeDistributed version as well. It's starting to be crowded in `core.py`. Thoughts on this?

I think we should look into a generic `TimeDistributed` layer that could time-distribute any existing layer. Now that we have the new backend,  and callable layers, that shouldn't be too difficult.

**Comment by lukedeo** [comment]
Hmm, I definitely agree on the crowdedness in `core.py`. Short term, I suppose we could either merge or create something like `timedistrib.py` (or something of that ilk), or hold off on the time distributed version of the `Highway` and work on a generic `TimeDistributed` layer. I suppose for that, we would just need a `step(x, states)` defined for each non time-distr. layer, then internally we can call `K.rnn(self.step, X, [], masking=False)` inside `TimeDistributed`, similar to what we have now. The only layer I can see this failing for is `TimeDistributedMerge`. Thoughts?

**Comment by fchollet** [comment]
> I suppose for that, we would just need a step(x, states) defined for each non time-distr. layer, then internally we can call K.rnn(self.step, X, [], masking=False) inside TimeDistributed

Couldn't we just have:
- a `TimeDistributed` layer that takes a layer as constructor argument
- a `step` function inside `TimeDistributed` that calls `self.layer` (which is now callable)
- `get_output` based on `rnn` (similar to `TimeDistributedDense`)

That way we don't need to modify existing non-time-distributed layers in any way.

Let's look into it. For now, I suggest that we merge the highway PR without the TimeDistributed version.

**Comment by lukedeo** [comment]
Cool, I just removed `TimeDistributedHighway` for the sake of the PR. I hope Travis isn't weird again...

I'll probably try to tackle this `TimeDistributed` layer business in the next few days. I use these guys a lot so some abstraction would be nice.

**Comment by fchollet** [comment]
Cool, thanks!
